### PR TITLE
Add validation for s3 bucket to be in same region of cluster stack

### PR DESCRIPTION
**Pull request by chenwany**
Add a validator to validate that `cluster_resource_bucket` is in the same region with the cluster stack.
Tests:
When specify `cluster_resource_bucket` in a different region with where the cluster create, the validator will generate the following error:
ERROR: The configuration parameter 'cluster_resource_bucket' generated the following errors:
cluster_resource_bucket must be in the same region of the cluster.

Signed-off-by: chenwany <chenwany@amazon.com>

**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/config/validators.py [hunk 1]
```diff
@@ -769,6 +769,41 @@ def s3_bucket_validator(param_key, param_value, pcluster_config):
...
+    """Validate S3 bucket  is in the same region with the cloudformation stack."""
+    errors = []
+    warnings = []
+    s3_client = boto3.client("s3")
+    try:
+        bucket_region = s3_client.get_bucket_location(Bucket=param_value).get("LocationConstraint")
+        # Buckets in Region us-east-1 have a LocationConstraint of null
+        if bucket_region is None:
+            bucket_region = "us-east-1"
+        if bucket_region != pcluster_config.region:
```
Comment (rexcsn) [inline cli/src/pcluster/config/validators.py:782]: We need to consider the case that the region is provided through the CLI, i.e. `pcluster create <some_cluster> -r us-east-1`. We might need to validate this in `commands.py` or somehow pass in the region argument

File: cli/tests/pcluster/config/test_validators.py [hunk 2]
```diff
@@ -545,6 +546,69 @@ def test_s3_validator(mocker, boto3_stubber, config, num_calls, bucket, expected
...
+            "AccessDenied",
+            "You do not have access to the S3 bucket 'access_denied_bucket'",
+        ),
+    ],
+)
+def test_s3_bucket_region_validator(mocker, boto3_stubber, error_code, bucket, expected_message):
+    os.environ["AWS_DEFAULT_REGION"] = "eu-west-1"
+    get_bucket_location_response = {
+        "ResponseMetadata": {},
+        "LocationConstraint": "af-south1",
```
Comment (rexcsn) [inline cli/tests/pcluster/config/test_validators.py:573]: `us-east-1` is an edge case, so we should add a test case for that.
We also need to add a test case where the validation passes for full coverage.

File: cli/tests/pcluster/config/test_validators.py [hunk 3]
```diff
@@ -545,6 +546,69 @@ def test_s3_validator(mocker, boto3_stubber, config, num_calls, bucket, expected
...
+        ),
+        (
+            "access_denied_bucket",
+            "AccessDenied",
+            "You do not have access to the S3 bucket 'access_denied_bucket'",
+        ),
+    ],
+)
+def test_s3_bucket_region_validator(mocker, boto3_stubber, error_code, bucket, expected_message):
+    os.environ["AWS_DEFAULT_REGION"] = "eu-west-1"
```
Comment (rexcsn) [inline cli/tests/pcluster/config/test_validators.py:570]: While this is ok, would be nice if we can test out the different cases when region is provided with CLI vs. provided in original config.
If that logic is already tested, you can also leave a comment in the code to point to an existing test case here

File: cli/src/pcluster/config/validators.py (hunk 1 above)
Comment (chenwany) [reply cli/src/pcluster/config/validators.py:784]: The case `-r us-east-1` is already considered here, since the pcluster_config in the validator stage already convert all the input arguments into the config file.

File: cli/src/pcluster/config/validators.py [hunk 4]
```diff
@@ -769,6 +769,41 @@ def s3_bucket_validator(param_key, param_value, pcluster_config):
     return errors, warnings
 
 
+def s3_bucket_region_validator(param_key, param_value, pcluster_config):
+    """Validate S3 bucket  is in the same region with the cloudformation stack."""
```
[the remark this answers is not in the document]
Comment (chenwany) [reply cli/src/pcluster/config/validators.py]: Yes, fixed!

File: cli/src/pcluster/config/validators.py (hunk 1 above)
Comment (chenwany) [reply cli/src/pcluster/config/validators.py:784]: The case -r us-east-1 is already considered here, since the pcluster_config in the validator stage already convert all the input arguments into the config file.

File: cli/tests/pcluster/config/test_validators.py (hunk 3 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_validators.py]: Done! Here we provide the region by using os.environ["AWS_DEFAULT_REGION"], it should be passed to the config, although we use it in many places, it seems that we don't have test for this, we may need to add this test later

File: cli/tests/pcluster/config/test_validators.py (hunk 2 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_validators.py]: Added. Also added explainations in the comments of the code

File: cli/src/pcluster/config/validators.py [hunk 5]
```diff
@@ -797,6 +797,43 @@ def s3_bucket_validator(param_key, param_value, pcluster_config):
...
+                "You do not have access to the S3 bucket '{0}': '{1}'".format(
+                    param_value, client_error.response.get("Error").get("Message")
+                )
+            )
+        else:
+            errors.append(
+                "Unexpected error when calling get_bucket_location on S3 bucket '{0}': '{1}'".format(
+                    param_value, client_error.response.get("Error").get("Message")
+                )
+            )
```
Comment (rexcsn) [inline cli/src/pcluster/config/validators.py:832]: You can probably call `_process_generic_s3_bucket_error` to avoid code duplication.
We will need to fix the error message `Unexpected error when calling get_bucket_location on S3 bucket` in `_process_generic_s3_bucket_error`, because it should be a generic error.

nit: we can also fix similar code duplications in `fsx_lustre_auto_import_validator
` below

File: cli/tests/pcluster/config/test_validators.py [hunk 6]
```diff
@@ -546,6 +547,98 @@ def test_s3_validator(mocker, boto3_stubber, config, num_calls, bucket, expected
     utils.assert_param_validator(mocker, config, expected_message)
 
 
+@pytest.mark.parametrize(
+    "bucket, region, error_code, expected_message, client_error",
+    [
+        (
+            "bucket",
+            "us-east-1",
```
Comment (rexcsn) [inline cli/tests/pcluster/config/test_validators.py:555]: I would add another test case for a region that passes the validator and is not `us-east-1`, since `us-east-1` is a special case, we also want to test that other "normal" region can pass the validator.

File: cli/src/pcluster/config/validators.py (hunk 5 above)
Comment (chenwany) [reply cli/src/pcluster/config/validators.py:832]: Done! Aslo fixed that for `fsx_lustre_auto_import_validator`

File: cli/tests/pcluster/config/test_validators.py (hunk 6 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_validators.py:555]: Added

**Review by rexcsn — DISMISSED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 ------
 **ENHANCEMENTS**
 
+- Add validation to prevent `cluster_resource_bucket` creating not in same region as cluster.
```
Comment (rexcsn) [inline CHANGELOG.md:8]: How about "Add validation to prevent using a `cluster_resource_bucket` that is in a different region than the cluster"

Comment (chenwany) [reply CHANGELOG.md:8]: Good idea, done!

**Review by rexcsn — APPROVED** [review]
